src/main/java/common/Storage.java

```diff
@@ -0,0 +1,105 @@
+
+    /**
+     * Constructor for Storage.
+     *
+     * @param storageName    Name of storage file. Should be a .txt file.
+     * @param storageDirName Name of directory storing the storage file.
+     */
+    public Storage(String storageName, String storageDirName) {
+        this.storageName = storageName;
+        this.storageDirname = storageDirName;
+        this.filePath = String.format("%s%s%s", this.storageDirname, File.separator, this.storageName);
+    }
+
+
+    /**
+     * Attempts to create a directory and/or storage file if it does not exist.
+     *
+     * @throws IOException Thrown when an error creating the storage file occurs.
+     */
+    public void initializeStorage() throws IOException {
+        Path filePath = Paths.get(this.filePath);
+        Path dirPath = Paths.get(storageDirname);
+
+        if (!Files.exists(dirPath)) {
+            Files.createDirectory(dirPath);
+            Ui.printCreateNewDirectory(this.storageDirname);
+        }
+        if (!Files.exists(filePath)) {
+            Files.createFile(filePath);
+            Ui.printCreateNewStorage(this.storageName);
+        }
+    }
+
+    /**
+     * Writes a given task list to storage, with the delimiter used being "#".
+     *
+     * @param taskList Task list to save to storage.
+     */
+    public void writeToStorage(TaskList taskList) {
+        try {
+            FileWriter fw = new FileWriter(this.filePath);
+            for (int i = 0; i < taskList.size(); i++) {
+                fw.write(taskList.get(i).getEncodedValue() + "\n");
+            }
+            fw.close();
+        } catch (IOException e) {
+            Ui.printError(e);
+        }
+    }
+
+    /**
+     * Reads from an existing storage file and populates an empty TaskList object with Task objects
+     * generated from storage data.
+     *
+     * @param taskList The TaskList object to populate.
+     * @throws IOException Thrown when reading errors occur.
+     */
+    public void readFromStorage(TaskList taskList) throws IOException {
+        File f = new File(this.filePath);
+        Scanner s = new Scanner(f); // create a Scanner using the File as the source
+        while (s.hasNext()) {
+            String[] encodedTask = s.nextLine().split("#");
+            switch (encodedTask[0]) {
+                case "[T]": {
+                    taskList.addTask(new ToDo(encodedTask[1], Boolean.parseBoolean(encodedTask[2])));
+                    break;
+                }
+                case "[D]": {
+                    taskList.addTask(new Deadline(encodedTask[1], Boolean.parseBoolean(encodedTask[2]), DeadlineCommand.parseDeadlineDatetimeFromStorage(encodedTask[3])));
+                    break;
+                }
+                case "[E]": {
+                    taskList.addTask(new Event(encodedTask[1], Boolean.parseBoolean(encodedTask[2]), EventCommand.parseEventDatetimeFromStorage(encodedTask[3]), EventCommand.parseEventDatetimeFromStorage(encodedTask[4])));
```

Comment:
This line here seems a bit too long, perhaps a line break?
```suggestion
                    taskList.addTask(new Event(encodedTask[1], 
                         Boolean.parseBoolean(encodedTask[2]), 
                         EventCommand.parseEventDatetimeFromStorage(encodedTask[3]), 
                         EventCommand.parseEventDatetimeFromStorage(encodedTask[4])));
```